timemachine/exchange/exchange_mover.py

```diff
@@ -87,108 +95,385 @@ def get_initial_state(water_pdb, mol, ff, seed, nb_cutoff):
+        chosen_water = np.random.choice(np.arange(self.num_waters), p=probs_before)
+        chosen_water_atoms = self.water_idxs_np[chosen_water]
 
         # compute delta_U of insertion
         trial_chosen_coords = coords[chosen_water_atoms]
         trial_translation = np.diag(box) * np.random.rand(3)
         # tbd - what should we do  with velocities?
 
-        # this has a higher acceptance probability than if we allowed for rotations
-        # (probably because we have a much higher chance of a useless move)
-        moved_coords = randomly_translate(trial_chosen_coords, trial_translation)
-        trial_coords = coords.copy()  # can optimize this later if needed
-        trial_coords[chosen_water_atoms] = moved_coords
+        moved_coords = randomly_rotate_and_translate(trial_chosen_coords, trial_translation)
+        # moved_coords = randomly_translate(trial_chosen_coords, trial_translation)
+
+        # optimized version using double transposition
+        log_weights_after, trial_coords = self.batch_log_weights_incremental(
+            coords, box, chosen_water, moved_coords, log_weights_before
+        )
+
+        # reference version
+        # trial_coords = coords.copy()  # can optimize this later if needed
+        # trial_coords[chosen_water_atoms] = moved_coords
+        # log_weights_after = self.batch_log_weights(trial_coords, box)
+
+        log_p_accept = min(0, logsumexp(log_weights_before) - logsumexp(log_weights_after))
+        new_state = CoordsVelBox(trial_coords, x.velocities, x.box)
+
+        return new_state, log_p_accept
+
 
-        delta_U_insert = self.U_fn(trial_coords, box, a_idxs, b_idxs)
-        # If our system is in a clash free state, a deletion move has delta_Us typically
-        # on the order of 35kTs. However, if we start in a clashy state, we can no longer
-        # guarantee this. So it's safer to disable this micro-optimization for now.
-        delta_U_delete = -self.U_fn(coords, box, a_idxs, b_idxs)
-        delta_U_total = delta_U_delete + delta_U_insert
+# numpy version of delta_r
+def delta_r_np(ri, rj, box):
+    diff = ri - rj  # this can be either N,N,3 or B,3
+    if box is not None:
+        box_diag = np.diag(box)
+        diff -= box_diag * np.floor(diff / box_diag + 0.5)
+    return diff
+
+
+def v1_insertion(radius, center, box):
+    # (ytz): sample a point inside the sphere uniformly
+    # source: https://karthikkaranth.me/blog/generating-random-points-in-a-sphere/
+    xyz = np.random.randn(3)
+    n = np.linalg.norm(xyz)
+    xyz = xyz / n
+    # (ytz): this might not be inside the box, but it doesnt' matter
+    # since all of our distances are computed using PBCs
+    c = np.cbrt(np.random.rand())
+    new_xyz = xyz * c * radius + center
+    assert np.linalg.norm(delta_r_np(new_xyz, center, box)) < radius
+
+    return new_xyz
+
+
+def v2_insertion(radius, center, box):
+    # generate random proposals outside of the sphere but inside the box
+    while True:
```

Comment:
preference: avoid possible infinite loop if radius is large relative to box -- perhaps a for loop with an error if this doesn't succeed after a large number of tries?

Reply:
good idea - willfix.